riscv/runtime/src/fmt.rs

```diff
@@ -1,22 +1,37 @@
+extern crate alloc;
+
 use core::arch::asm;
 use core::fmt;
 use core::mem::MaybeUninit;
 
-// #[macro_export]
-// macro_rules! print {
-//     ($($arg:tt)+) => (print_args(format_args!( $($arg)+)))
-// }
+#[macro_export]
+macro_rules! print {
+    ($($arg:tt)+) => {{
+        $crate::fmt::print_args(format_args!( $($arg)+));
+    }};
+}
+
+pub fn print_args(args: fmt::Arguments) {
+    print_str(&alloc::fmt::format(args));
+}
+
+/// Like print, but doesn't allocate. Instead, truncates to the size (in bytes)
+/// provided as first argument (must be constant).
+#[macro_export]
+macro_rules! print_truncate {
+    ($len:expr, $($arg:tt)+) => {{
+        $crate::fmt::print_truncate_args::<{ $len }>(format_args!( $($arg)+));
+    }};
+}
 
-// TODO turn this into a macro
-pub fn print(args: fmt::Arguments) {
-    const BUF_SIZE: usize = 1024;
-    let mut buf = unsafe { MaybeUninit::<[MaybeUninit<u8>; BUF_SIZE]>::uninit().assume_init() };
+pub fn print_truncate_args<const BUF_SIZE: usize>(args: fmt::Arguments) {
+    let mut buf = [MaybeUninit::<u8>::uninit(); BUF_SIZE];
     let _s: &str = buf_formatter::format(&mut buf, args).unwrap();
-    print_prover(_s);
+    print_str(_s);
 }
 
-#[inline]
-fn print_prover(s: &str) {
+// Please don't let this function panic...
```

Comment:
What do you want to say with this comment? Is there a way this could panic? What happens if it does?

Reply:
This is to the next person who touches this code. If it panics, panic handler will call itself recursively until stack overflows.

Reply:
Sure, but then please write that into the comment.